src/peakrdl_sv/regmodel.py

```diff
@@ -22,8 +52,357 @@ def _parse_rdl(self, rdlfile: str) -> AddressMap:
+        :type reg_name: str
+        :return: The target register, None if the register name is not found in the
+        regmap
+        :rtype: Register | None
+        """
+        try:
+            return self._reg_map[reg_name]
+        except KeyError:
+            self._log.info(f"Could not find register in {self._reg_map.keys()}")
+            return None
+
+    def split_value_over_fields(
+        self,
+        target: Register,
+        value: int,
+    ) -> Generator[tuple[Any, int], Any, None]:
+        """Split a value over the fields of a register with correct masking depending
+        on the field location
+
+        :param target: The register to target
+        :type target: Register
+        :param value: The value to write to the registers
+        :type value: int
+        :yield: tuples of the field address and the value to write
+        :rtype: Generator[tuple[Any, int], Any, None]
+        """
+
+        # format the write value an `regwidth`-bit binary string
+        bits = f"{value:0{target.regwidth}b}"
+
+        # Extract the bits we are going to write and conver them to int
+        for field in target:
+            # get the slice indices of the field
+            lower = target.regwidth - field.msb - 1
+            upper = target.regwidth - field.lsb
+
+            # perform the slice (python slicing non inclusive)
+            yield (field.absolute_address, int(bits[lower:upper], 2))  # noqa: E203
+
+    # --------------------------------------------------------------------------------
+    # Generic read and write wrappers
+    # --------------------------------------------------------------------------------
+
+    async def write(
+        self,
+        reg_name: str,
+        field_name: str | None = None,
+        data: int | None = None,
+        **kwargs,
+    ) -> None:
+        """Writes the value of the DUT register. Two options for calling:
+
+        (1) register name + data -> write to named register a particular value
+        (2) register name -> write desired value to a named register
+
+        :param reg_name: Register name to write to
+        :type reg_name: str | int
+        :param data: Data to write to register, defaults to None
+        :type data: int | None, optional
+        """
+
+        if isinstance(field_name, int):
+            self._log.warning(
+                "You passed an int to the `field_name` argument of write(), this is"
+                " probably a mistake, try calling with data=`value`",
+            )
+
+        async def write_desired_to_named_reg(reg_name: str):
+            """Writes a desired register value to the DUT"""
+
+            target = self.get_register_by_name(reg_name)
+
+            # write the register value as multiple `accesswidth` chunks
+            if target.regwidth > target.accesswidth:
+                for field in target:
+                    await self._callbacks.async_write_callback(
```

Comment:
I don't understand what's going on here.  I think perhaps you're assuming that a field has the same width as `accesswidth`.

Reply:
Yes that was my assumption, and I see that it is incorrect.What I am now realising is that for all cases (regardless of 'wideness') if the field's LSB is not alligned to the access width then we need to shift the desired by `field.lsb` to the left with consideration for the fields offset in the register. Something like

```
def align_to_field(self, target: Field, value: int) -> int:
    resolved_lsb = target.lsb % target.parent.accesswidth
    return value << resolved_lsb
```